spec/support/fake_analytics.rb

```diff
@@ -102,22 +105,51 @@ def browser_attributes
     if attributes_matcher.nil?
       expect(actual.events).to have_key(event)
     else
-      expect(actual.events[event]).to include(match(attributes_matcher))
+      attributes_for_events = actual.events[event].pluck(:attributes)
+      expect(attributes_for_events).to include(match(attributes_matcher))
     end
+
+    if !allow_multiple_events?
+      expect(actual.events[event]).to have_attributes(size: 1)
+    end
+  end
+
+  chain :at_least_once do
+    @allow_multiple_events = true
   end
 
   failure_message do |actual|
     matching_events = actual.events[event]
-    if matching_events&.length == 1 && attributes_matcher.instance_of?(Hash)
+
+    if matching_events&.length.to_i > 1 && !allow_multiple_events?
+      <<~MESSAGE
+        FakeAnalytics received too many #{event} events.
+        expected: 1
+        got:      #{matching_events.length}
+
+        Events received:
+
+        #{matching_events.map do |event| 
+          <<~EVENT
+            at #{first_relevant_backtrace_line(event[:backtrace])}:
+            #{event[:attributes].pretty_inspect.split("\n").map { |line| "  #{line}" }.join("\n")},
+          EVENT
+        end.
+        join("\n\n")}
```

Comment:
personally I like to keep these on the same line
```suggestion
        end.join("\n\n")}
```